Refactored SDK command to use downloads.titaniumsdk.com

cb1kenobi opened the pull request:
Fixes #585

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -369,91 +319,74 @@ SdkSubcommands.select = {
 };
 
+function sortVersion(a, b) {
+	const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(versionRegExp);
```
No need to have the `_` here unless you're doing it to make it obvious
```suggestion
	const [ , amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(versionRegExp);
```

cb1kenobi replied:
Fixed.

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -369,91 +319,74 @@ SdkSubcommands.select = {
 
+function sortVersion(a, b) {
+	const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(versionRegExp);
+	const [ __, bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(versionRegExp);
```
Same here
```suggestion
	const [ , bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(versionRegExp);
```

cb1kenobi replied:
Fixed

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -535,713 +473,575 @@ async function select(logger, config, cli) {
-	} catch (error) {
-		logger.error(error);
-		throw error;
+	const titaniumDir = afs.resolvePath(cli.env.installPath);
```
You've spaced the others nicely, do that here too if you're wanting to do that?

cb1kenobi replied:
Fixed.

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -535,713 +473,575 @@ async function select(logger, config, cli) {
+		// step 1.5: download the file
 
-/**
- * Validates the zip that has been requested to be installed. This is because when downloading a
```
I think it's good to keep the context of why we look inside the zip for some things

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
+async function getBranches() {
+	return Object
+		.entries((await got('https://downloads.titaniumsdk.com/registry/branches.json', { responseType: 'json' })).body)
+		.filter(([ name, count ]) => count)
```
```suggestion
		.filter(([, count ]) => count)
```

cb1kenobi replied:
Fixed.

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
+	}).filter(r => {
+		return r.assets.some(a => a.os === os);
+	}).sort((a, b) => {
+		const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(releaseRE);
```
```suggestion
		const [ , amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(releaseRE);
```

cb1kenobi replied:
Fixed.

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -1406,179 +1206,78 @@ async function uninstall(logger, config, cli) {
+		return r.assets.some(a => a.os === os);
+	}).sort((a, b) => {
+		const [ _, amajor, aminor, apatch, atag ] = a.name.toLowerCase().match(releaseRE);
+		const [ __, bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(releaseRE);
```
```suggestion
		const [ , bmajor, bminor, bpatch, btag ] = b.name.toLowerCase().match(releaseRE);
```

cb1kenobi replied:
Fixed.

ewanharris reviewed (commented):
Some comments from a first pass, need to do a functional test

ewanharris commented:
@cb1kenobi what're your plans for the existing tests for this?

m1ga commented:
`ti sdk select 10.1.1.GA` (all select commands) will throw:
```
[ERROR] TypeError: Cannot read properties of undefined (reading 'toLowerCase')
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:323:53)
    at Array.sort (<anonymous>)
    at Function.select (/home/miga/tools/titanium/lib/commands/sdk.js:358:4)
    at Function.selectCallbackified (node:util:304:5)
    at run (/home/miga/tools/titanium/lib/commands/sdk.js:75:29)
    at CLI.<anonymous> (/home/miga/tools/titanium/lib/cli.js:808:5)
    at wrapper (/home/miga/tools/titanium/node_modules/async/dist/async.js:271:20)
    at Object.iteratorCallback (/home/miga/tools/titanium/node_modules/async/dist/async.js:501:17)
    at Object.<anonymous> (/home/miga/tools/titanium/node_modules/async/dist/async.js:327:20)
    at CLI._fireHookCallback (/home/miga/tools/titanium/lib/hook.js:274:12)
```

a and b don't have a `name` in my tests, its only the string

m1ga commented:
minor issue: 
`titanium sdk install 11.0.0.RC` will always download the file and after that says `already installed`. The current version checks it before downloading.
With `--force` it worked fine :+1:

m1ga commented:
If you have a master build folder like `11.0.0` the `ti sdk select` regex will fail: https://github.com/tidev/titanium/blob/96d4881ad5a0badcb48c8936bcc152338cc84cab/lib/commands/sdk.js#L321
My `a` and `b` are: `11.1.0 9.3.2.GA` and it won't match `11.1.0`

```
[ERROR] TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:321:45)
    at Array.sort (<anonymous>)
    at Function.select (/home/miga/tools/titanium/lib/commands/sdk.js:356:4)
    at Function.selectCallbackified (node:util:304:5)
    at run (/home/miga/tools/titanium/lib/commands/sdk.js:75:29)
    at CLI.<anonymous> (/home/miga/tools/titanium/lib/cli.js:808:5)
    at wrapper (/home/miga/tools/titanium/node_modules/async/dist/async.js:271:20)
    at Object.iteratorCallback (/home/miga/tools/titanium/node_modules/async/dist/async.js:501:17)
    at Object.<anonymous> (/home/miga/tools/titanium/node_modules/async/dist/async.js:327:20)
    at CLI._fireHookCallback (/home/miga/tools/titanium/lib/hook.js:274:12)
```

cb1kenobi commented:
@m1ga Fixed the version regexp.

m1ga commented:
Regex is better now :+1:  

It still failed for me but I had some folder like `10.2.0.Fix` but I did that by hand. Old CLI didn't care, but again: it's just my SDK folder. But we could add a warning like "unsupported SDK found in SDK folder" instead of 
```
[ERROR] TypeError: object null is not iterable (cannot read property Symbol(Symbol.iterator))
    at sortVersion (/home/miga/tools/titanium/lib/commands/sdk.js:321:45)
   ...
```
if there is a problem there. But nothing for this release!

---

**Tested:**
* list, select (valid and invalid, version and by list) :heavy_check_mark: 
* install valid version, latest, force, ci build  :heavy_check_mark: 
* install invalid is working and shows this error  :heavy_check_mark: 
```
[ERROR] Error: Unable to find any Titanium SDK releases or CI builds that match "9.4.2"
    at Function.install (/home/miga/tools/titanium/lib/commands/sdk.js:635:10)
```
old SDK only showed it as a "log.error" so without the sdk.js line. But I'm fine with this too. Just don't want people to think its an error in the file/function!

---

_Minor_ cosmetic issues that can be changed later (if at all):
* installing a SDK that you already have will start the download and then say: its already installed use force. When you do that it will download it again. So it would be nice (like the current CLI does it) not downloading it at all and say "use force".
* installing a CLI build will just show the spinner and no progress bar. At first I thought it was stuck. If you don't get a progress for nightly builds we could just add a `.` every two second. Might fill up the screen for slower network speeds but at least it gives the impression that it is working.
* warning for unsupported SDKs [see above]

cb1kenobi commented:
@m1ga I've addressed most of your concerns: fixed version regex (again), added overwrite prompt during install, and don't show stack trace for known error messages.

caspahouzer reviewed (approved):
Same here. Works great. 

LGTM

cb1kenobi commented on lib/commands/sdk.js:
```diff
@@ -262,86 +225,71 @@ async function list(logger, config, cli) {
+				logger.log(`   ${r.name.padEnd(12).cyan}\
+${Intl.DateTimeFormat('en-US', { dateStyle: 'short' }).format(new Date(r.date)).padStart(8)}\
+${humanize.filesize(r.assets.find(a => a.os === os).size, 1024, 1).toUpperCase().padStart(11)}\
+${Object.hasOwn(sdks, r) ? ' [installed]' : ''}\
```
Fixed

ewanharris commented on lib/commands/sdk.js:
```diff
@@ -175,83 +159,64 @@ async function list(logger, config, cli) {
+
+		const obj = {
 			activeSDK,
+			branch: branchBuilds.length ? {
```
`branchBuilds` can be undefined here so need to guard against that
```suggestion
			branch: branchBuilds?.length ? {
```

cb1kenobi replied:
Fixed

ewanharris reviewed (changes requested):
The output of `ti sdk list --release --output json` has changed which is going to break the editors. Either we'll need to patch those ahead of this release shipping or we could massage the output?

Comparison of the two outputs [here](https://gist.github.com/ewanharris/08e9bc0e9e7b4bb524a5bb80ebba7d5d)

m1ga commented:
> The output of `ti sdk list --release --output json` has changed which is going to cause issues for the editors. Either we'll need to patch those ahead of this release shipping or we could massage the output?

Good catch @ewanharris  :+1:

cb1kenobi commented:
Fixed the JSON output of the releases and branches. :)

ewanharris commented:
The RC inclusion I think will be an issue too, as the [logic used currently](https://github.com/tidev/titanium-editor-commons/blob/56f3338cd0ed36fa7717039af702248e824eaa8e/src/updates/titanium/sdk.ts#L57) just handles GA.

I can update the editors to handle RC (just ignore it for now) to avoid going round on this PR again, would be best to wait a few days for that to roll out I think

cb1kenobi commented:
I could add new `--beta` and `--rc` options to the list command to include those in the list. What do you think?

ewanharris commented:
That would be a nice addition that would probably quicker than releasing the editors. Maybe combine under a single `--preview` flag that would toggle both rather than flags for each?

cb1kenobi commented:
I don't love `--preview`. I could get behind `--unstable`. Thoughts?

cb1kenobi commented:
Added the `--unstable` flag.
